src/main/java/com/semmle/jira/addon/config/init/WorkflowUtils.java

```diff
@@ -0,0 +1,116 @@
+    if (workflowManager.workflowExists(workflowName)) {
+      return;
+    }
+
+    Map<String, WorkflowStatus> statuses = mapStatuses(statusesJson);
+
+    WorkflowDescriptor descriptor = WorkflowUtil.convertXMLtoWorkflowDescriptor(workflowXml);
+    List<StepDescriptor> steps = descriptor.getSteps();
+    for (StepDescriptor step : steps) {
+      Map<String, String> meta = step.getMetaAttributes();
+      String statusName = step.getName();
+      WorkflowStatus workflowStatus = statuses.get(statusName);
+
+      Status status = null;
+      // statusesJson does not include all status
+      // not included statuses should be default ones (Open, Closed)
+      if (workflowStatus == null) {
+        status = getOrCreateStatus(statusName, "", 0);
+      } else {
+        status =
+            getOrCreateStatus(
+                statusName,
+                workflowStatus.description,
+                Long.parseLong(workflowStatus.statusCategoryId));
+      }
+      meta.put("jira.status.id", status.getId());
+    }
+
+    JiraWorkflow workflow = new ConfigurableJiraWorkflow(workflowName, descriptor, workflowManager);
+    workflowManager.createWorkflow(
+        (ApplicationUser) null, workflow); // User is not needed for creating
+  }
+
+  static void addLayoutToWorkflow(String workflowName, String layoutJson) {
+    ComponentAccessor.getComponent(OfBizDelegator.class);
+    OfBizDelegator ofBizDelegator = ComponentAccessor.getComponent(OfBizDelegator.class);
+    String layoutKey = DigestUtils.md5Hex(workflowName);
+    final Map<String, Object> entryFields =
+        ImmutableMap.of(
+            "entityName",
+            "com.atlassian.jira.plugins.jira-workflow-designer",
+            "entityId",
+            1,
+            "propertyKey",
+            "jira.workflow.layout.v5:" + layoutKey,
+            "type",
+            6);
+    final Long layoutId = ofBizDelegator.createValue("OSPropertyEntry", entryFields).getLong("id");
+    final Map<String, Object> textFields =
+        ImmutableMap.of(
+            "id", layoutId,
+            "value", layoutJson);
+    ofBizDelegator.createValue("OSPropertyText", textFields);
+  }
+
+  private static Status getOrCreateStatus(
+      String statusName, String description, long statusCategoryId) {
+    Status status =
+        (Status)
+            ComponentAccessor.getConstantsManager()
+                .getIssueConstantByName(
+                    ConstantsManager.CONSTANT_TYPE.STATUS.getType(), statusName);
+    if (status != null) {
+      return status;
+    }
+
+    StatusManager statusManager = ComponentAccessor.getComponent(StatusManager.class);
+    StatusCategory category = StatusCategoryImpl.findById(statusCategoryId);
+
+    return statusManager.createStatus(statusName, description, "status.png", category);
+  }
+
+  private static Map<String, WorkflowStatus> mapStatuses(String statusesJson) {
+    Type collectionType = new TypeToken<Collection<WorkflowStatus>>() {}.getType();
```

Comment:
You might want to use an array type here `WorkflowStatus[]` . That way you can avoid the arcane TypeToken stuff.

```
WorkflowStatus[] statuses = new Gson().fromJson(statusesJson, WorkflowStatus[].class);
```